File: backend/app/schemas/compare.py
"""对比 Agent 请求/响应 Schema"""
from datetime import datetime

from pydantic import BaseModel, Field


# ── 创建会话 ──────────────────────────────────────────────────────

class CompareSessionCreate(BaseModel):
    property_ids: list[int] = Field(..., min_length=2, max_length=10)
    priority: str = Field("balanced", pattern=r"^(balanced|budget|commute|space|safety)$")


class CompareSessionResponse(BaseModel):
    id: int
    user_id: int
    property_ids: list[int]
    priority: str
    status: str
    result_cache: dict | None = None
    created_at: datetime
    messages: list["CompareMessageRead"] = []

    model_config = {"from_attributes": True}


class CompareMessageRead(BaseModel):
    id: int
    role: str
    content: str | None = None
    tool_calls: dict | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── 发送消息 ──────────────────────────────────────────────────────

class CompareMessageRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=2000)
    priority: str | None = Field(None, pattern=r"^(balanced|budget|commute|space|safety)?$")


class CompareMessageResponse(BaseModel):
    reply: str
    scores: dict[int, dict] = {}         # {property_id: {total: int, breakdown: {dim: int}}}
    tool_trail: list[dict] = []          # 调试/审计：工具调用轨迹
    property_data: dict[int, dict] = {}  # {property_id: EnrichedPropertyData}

File: frontend/src/types/compare.ts
/** 对比 Agent 类型定义 */

export type ComparePriority = 'balanced' | 'budget' | 'commute' | 'space' | 'safety'

export interface CompareSessionCreate {
  property_ids: number[]
  priority?: ComparePriority
}

export interface CompareSessionResponse {
  id: number
  user_id: number
  property_ids: number[]
  priority: ComparePriority
  status: 'active' | 'completed'
  result_cache: CompareResultCache | null
  created_at: string
  messages: CompareMessage[]
}

export interface CompareResultCache {
  scores: Record<number, DimensionScores>
  property_data: Record<number, EnrichedPropertyData>
  reply: string
}

export interface CompareMessage {
  id: number
  role: 'user' | 'assistant' | 'tool'
  content: string | null
  tool_calls: Record<string, unknown> | null
  created_at: string
}

export interface CompareMessageRequest {
  message: string
  priority?: ComparePriority | null
}

export interface CompareMessageResponse {
  reply: string
  scores: Record<number, DimensionScores>
  tool_trail: ToolCallRecord[]
  property_data: Record<number, EnrichedPropertyData>
}

export interface DimensionScores {
  total: number
  breakdown: Record<string, number>
}

export interface EnrichedPropertyData {
  property_id: number
  title: string
  district: string
  price_monthly: number
  area_sqm: number | null
  bedrooms: number
  bathrooms: number
  property_type: string
  amenities: string[]
  deposit_amount: number | null
  deposit_type: string | null
  service_fee_rate: number | null
  min_lease_months: number
  floor: number | null
  image_count: number
  transit_display: string | null
  rating: number | null
  review_count: number
  safety_score: number | null
}

export interface ToolCallRecord {
  tool: string
  arguments: Record<string, unknown>
  result: string
}
